slf4j/README.md

```diff
@@ -3,11 +3,26 @@ SLF4J
 
 This module allows directing Feign's logging to [SLF4J](http://www.slf4j.org/), allowing you to easily use a logging backend of your choice (Logback, Log4J, etc.)
 
-To use SLF4J with Feign, add both the SLF4J module and an SLF4J binding of your choice to your classpath.  Then, configure Feign to use the Slf4jLogger:
+To use SLF4J with Feign, add both the SLF4J module and an SLF4J binding of your choice to your classpath.  Then, configure Feign to use the Slf4jLogger.
+
+## Logger implementation
 
 ```java
 GitHub github = Feign.builder()
                      .logger(new Slf4jLogger())
                      .logLevel(Level.FULL)
                      .target(GitHub.class, "https://api.github.com");
 ```
+
+## Logger capability
+
+```java
+GitHub github = Feign.builder()
+                     .addCapability(
+                         Slf4jCapability.builder()
+                                        .withHeaders()
+                                        .withRequest()
+                                        .withResponse()
```

Comment:
Yeah, there is small mess with naming:
* withHeaders means "with request and response headers"
* withResponse means "with response body" and someone can  be fooled that it means "with response headers and body".

Maybe I need next builder methods:
* withHeaders - request+response headers
* withBody - request+response bodies
* withRequest - request headers and body
* withResponse - response headers and body

Reply:
I recommend that instead of a builder with a lot of options controlling the output, we consider allowing users to provide callback functions, allowing the users to decide what's included and what's not.

Having spent a long time in this area recently, I've come to find that folks have many different requirements for what they allow to appear in their logs.  Many security organizations require sanitization.  Others require use of MDC/ThreadContext.  With type of update, we have an opportunity to refactor this capability to allow for this level of customization, while still controlling the checkpoints in request processing.

Thoughts?

Reply:
My current implementation is too hard to extend. So first step: split interfaces LoggerCapability, LoggerConfiguration and abstract classes LoggedClient and LoggedRetryer from the implementation.

Now we have list of logging steps(methods):
* log basic request data (URL),
* log request headers,
* log request body,
* log basic response data (status and execution time),
* log response headers,
* log response body,
* two special steps for exception and retriyng call.

Maybe we need to add something for initialization and another finalizing method.